### ISSUE TITLE
determine why musl binaries are much slower than glibc binaries especially when threading is used

### ISSUE THREAD

**Issue by radhermit** [pkgcraft/pkgcraft#179]
One example of this would be running full repo scans using pkgcruft with both binary types on the same system. For some configurations the glibc-based binary will be almost 50x faster than the musl-based one.

**Comment by thesamesam** [comment]
My best guess would be musl's malloc. Try shoving in mimalloc or jemalloc and see if it gets better?

**Comment by radhermit** [comment]
Good point, I recall reading issues relating to rust binary optimization and allocator usage years ago and so partially thought more of that would be fixed without me having to do anything by now. :laughing:

**Comment by radhermit** [comment]
Adding a semi-recent, relevant blog post related to mimalloc for reference: https://www.tweag.io/blog/2023-08-10-rust-static-link-with-mimalloc/

**Comment by radhermit** [comment]
As an update, initial test builds using mimalloc on the musl branch yield no speedups and are still spending a majority of time in kernel space vs glibc builds (as seen in htop). I expect the same result from using jemalloc instead.

It's probably worth stracing the related memory allocation syscalls (sbrk, brk, and/or mmap) to see if anything can be divined from them.

**Comment by lu-zero** [comment]
https://docs.rs/bumpalo/latest/bumpalo/ might help but I'd optimize this once the other pending refactors land since it is very invasive.
